File: day20/main1.py
import os
from collections import deque

def read_grid(file_path):
    """
    Reads the racetrack grid from the input file.

    Args:
        file_path (str): Path to the input file.

    Returns:
        list: 2D list representing the grid.
        tuple: Start position (x, y).
        tuple: End position (x, y).
    """
    grid = []
    start = end = None
    with open(file_path, 'r') as f:
        for y, line in enumerate(f):
            row = list(line.strip())
            grid.append(row)
            if 'S' in row:
                start = (row.index('S'), y)
            if 'E' in row:
                end = (row.index('E'), y)
    return grid, start, end


def bfs(grid, start, end, allow_cheat=False):
    """
    Performs BFS to find the shortest path with optional cheating.

    Args:
        grid (list): 2D list representing the grid.
        start (tuple): Start position (x, y).
        end (tuple): End position (x, y).
        allow_cheat (bool): Whether to allow cheating.

    Returns:
        int: Shortest time to reach the end.
    """
    width, height = len(grid[0]), len(grid)
    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
    queue = deque([(start[0], start[1], 0, 0)])  # x, y, cost, cheats_used
    visited = set()

    while queue:
        x, y, cost, cheats_used = queue.popleft()

        if (x, y, cheats_used) in visited:
            continue
        visited.add((x, y, cheats_used))

        if (x, y) == end:
            return cost

        for dx, dy in directions:
            nx, ny = x + dx, y + dy
            if 0 <= nx < width and 0 <= ny < height:
                if grid[ny][nx] == '.' or grid[ny][nx] in 'SE':
                    queue.append((nx, ny, cost + 1, cheats_used))
                elif grid[ny][nx] == '#' and allow_cheat and cheats_used < 2:
                    queue.append((nx, ny, cost + 1, cheats_used + 1))

    return float('inf')


def count_cheats(grid, start, end, time_without_cheat):
    """
    Counts the number of cheats that save at least 100 picoseconds.

    Args:
        grid (list): 2D list representing the grid.
        start (tuple): Start position (x, y).
        end (tuple): End position (x, y).
        time_without_cheat (int): Time without any cheat.

    Returns:
        int: Number of cheats saving at least 100 picoseconds.
    """
    width, height = len(grid[0]), len(grid)
    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
    cheats_count = 0

    for y in range(height):
        for x in range(width):
            if grid[y][x] == '#':
                # Temporarily make this cell walkable to simulate a cheat
                grid[y][x] = '.'
                time_with_cheat = bfs(grid, start, end, allow_cheat=True)
                grid[y][x] = '#'  # Restore the cell

                if time_without_cheat - time_with_cheat >= 1:
                    #print(time_with_cheat)
                    cheats_count += 1

    return cheats_count


def main():
    """Main function to execute the solution."""
    # Define the input file path
    directory = os.path.dirname(os.path.abspath(__file__))
    file_path = os.path.join(directory, 'input.txt')
    file_path = os.path.join(directory, 'input2.txt')
    grid, start, end = read_grid(file_path)

    # Find shortest time without any cheat
    time_without_cheat = bfs(grid, start, end, allow_cheat=False)
    print(f"Time without cheat: {time_without_cheat}")

    # Count the cheats that save at least 100 picoseconds
    cheats_count = count_cheats(grid, start, end, time_without_cheat)
    print(f"Number of cheats saving at least 100 picoseconds: {cheats_count}")


if __name__ == "__main__":
    main()

File: day20/main1.go
package main

import (
	"bufio"
	"container/heap"
	"fmt"
	"log"
	"math"
	"os"
)

type Cell struct {
	X, Y int
}

type State struct {
	Cost       int
	X, Y       int
	CheatUsed  bool
	CheatEnded bool
	CSX, CSY   *int
	CEX, CEY   *int
}

type PriorityQueue []*State

func (pq PriorityQueue) Len() int { return len(pq) }
func (pq PriorityQueue) Less(i, j int) bool {
	return pq[i].Cost < pq[j].Cost
}
func (pq PriorityQueue) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
}
func (pq *PriorityQueue) Push(x interface{}) {
	*pq = append(*pq, x.(*State))
}
func (pq *PriorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)
	item := old[n-1]
	*pq = old[0 : n-1]
	return item
}

func readGrid(filePath string) ([][]rune, Cell, Cell) {
	file, err := os.Open(filePath)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	var grid [][]rune
	var start, end Cell
	y := 0
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		row := []rune(line)
		grid = append(grid, row)
		for x, ch := range row {
			if ch == 'S' {
				start = Cell{x, y}
			} else if ch == 'E' {
				end = Cell{x, y}
			}
		}
		y++
	}
	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}
	return grid, start, end
}

func bfsNoCheat(grid [][]rune, start, end Cell) int {
	width := len(grid[0])
	height := len(grid)
	directions := []Cell{{0, 1}, {1, 0}, {0, -1}, {-1, 0}}

	visited := make(map[Cell]bool)
	queue := []struct{ X, Y, Dist int }{{start.X, start.Y, 0}}

	for len(queue) > 0 {
		front := queue[0]
		queue = queue[1:]
		if front.X == end.X && front.Y == end.Y {
			return front.Dist
		}
		c := Cell{front.X, front.Y}
		if visited[c] {
			continue
		}
		visited[c] = true
		for _, d := range directions {
			nx, ny := front.X+d.X, front.Y+d.Y
			if nx < 0 || nx >= width || ny < 0 || ny >= height {
				continue
			}
			ch := grid[ny][nx]
			if ch == '.' || ch == 'S' || ch == 'E' {
				queue = append(queue, struct{ X, Y, Dist int }{nx, ny, front.Dist + 1})
			}
		}
	}
	return math.MaxInt32
}

func dijkstraWithOneCheatStep(grid [][]rune, start, end Cell, baseCost, minSavings int) (map[[2]Cell]int, int) {
	width := len(grid[0])
	height := len(grid)
	directions := []Cell{{0, 1}, {1, 0}, {0, -1}, {-1, 0}}

	// State: cost, x, y, cheatUsed, cheatEnded, c_sx, c_sy, c_ex, c_ey
	// We'll store visited as a map with a key:
	// key: (x, y, cheatUsed, cheatEnded, c_sx, c_sy, c_ex, c_ey)
	// However, c_ex, c_ey can be nil until cheat ended, same for c_sx, c_sy.
	// We'll store positions as integers:
	// If c_sx == nil, store as -1. Similarly for others.

	toInt := func(val *int) int {
		if val == nil {
			return -1
		}
		return *val
	}

	startState := &State{
		Cost: 0, X: start.X, Y: start.Y,
		CheatUsed: false, CheatEnded: false,
		CSX: nil, CSY: nil, CEX: nil, CEY: nil,
	}

	pq := &PriorityQueue{}
	heap.Init(pq)
	heap.Push(pq, startState)

	visited := make(map[[8]int]bool)
	cheatScenarios := make(map[[2]Cell]int)

	popCount := 0
	for pq.Len() > 0 {
		s := heap.Pop(pq).(*State)
		popCount++
		if popCount%100000 == 0 {
			fmt.Printf("Processed %d states...\n", popCount)
		}

		key := [8]int{s.X, s.Y, btoi(s.CheatUsed), btoi(s.CheatEnded), toInt(s.CSX), toInt(s.CSY), toInt(s.CEX), toInt(s.CEY)}
		if visited[key] {
			continue
		}
		visited[key] = true

		if s.X == end.X && s.Y == end.Y {
			// Check scenario validity
			if s.CheatUsed && s.CheatEnded && s.CSX != nil && s.CSY != nil && s.CEX != nil && s.CEY != nil && s.Cost < baseCost {
				timeSaved := baseCost - s.Cost
				scenarioKey := [2]Cell{{*s.CSX, *s.CSY}, {*s.CEX, *s.CEY}}
				if best, ok := cheatScenarios[scenarioKey]; !ok || best < timeSaved {
					cheatScenarios[scenarioKey] = timeSaved
				}
			}
			continue
		}

		for _, d := range directions {
			nx, ny := s.X+d.X, s.Y+d.Y
			if nx < 0 || nx >= width || ny < 0 || ny >= height {
				continue
			}

			ch := grid[ny][nx]
			newCost := s.Cost + 1

			if ch == '.' || ch == 'S' || ch == 'E' {
				// Track cell
				if s.CheatUsed && !s.CheatEnded {
					// Ending cheat now
					cx, cy := nx, ny
					ns := &State{
						Cost: newCost,
						X:    nx, Y: ny,
						CheatUsed:  true,
						CheatEnded: true,
						CSX:        s.CSX, CSY: s.CSY,
						CEX: &cx, CEY: &cy,
					}
					heap.Push(pq, ns)
				} else {
					ns := &State{
						Cost: newCost,
						X:    nx, Y: ny,
						CheatUsed:  s.CheatUsed,
						CheatEnded: s.CheatEnded,
						CSX:        s.CSX, CSY: s.CSY,
						CEX: s.CEX, CEY: s.CEY,
					}
					heap.Push(pq, ns)
				}
			} else {
				// Wall cell
				if s.CheatEnded {
					// Can't cheat anymore
					continue
				}
				if !s.CheatUsed {
					// Use single cheat step here
					sx, sy := s.X, s.Y
					ns := &State{
						Cost: newCost,
						X:    nx, Y: ny,
						CheatUsed:  true,
						CheatEnded: false,
						CSX:        &sx, CSY: &sy,
						CEX: nil, CEY: nil,
					}
					heap.Push(pq, ns)
				} else {
					// Already used cheat once, cannot do again
					continue
				}
			}
		}
	}

	count100 := 0
	for _, v := range cheatScenarios {
		if v >= minSavings {
			count100++
		}
	}

	return cheatScenarios, count100
}

func btoi(b bool) int {
	if b {
		return 1
	}
	return 0
}

func main() {
	filePath := "C:\\Programming\\Personal Projects\\AdventOfCode2024\\day20\\input.txt"
	//filePath := "C:\\Programming\\Personal Projects\\AdventOfCode2024\\day20\\input2.txt"
	grid, start, end := readGrid(filePath)

	timeWithoutCheat := bfsNoCheat(grid, start, end)
	fmt.Printf("Time without cheats: %d\n", timeWithoutCheat)

	cheatScenarios, count100 := dijkstraWithOneCheatStep(grid, start, end, timeWithoutCheat, 100)

	savingsCount := make(map[int]int)
	for _, ts := range cheatScenarios {
		savingsCount[ts]++
	}

	fmt.Println("All Cheats by Time Saved:")
	keys := make([]int, 0, len(savingsCount))
	for k := range savingsCount {
		keys = append(keys, k)
	}
	// Sort keys
	for i := 0; i < len(keys)-1; i++ {
		for j := i + 1; j < len(keys); j++ {
			if keys[i] > keys[j] {
				keys[i], keys[j] = keys[j], keys[i]
			}
		}
	}
	for _, k := range keys {
		fmt.Printf("Time saved: %d, Count: %d\n", k, savingsCount[k])
	}
	fmt.Printf("\nNumber of cheats that save at least 100 picoseconds: %d\n", count100)
}
